File: libraries/hds/dat_ccopy.f
*+ DAT_CCOPY - Copy one level of structure
      subroutine dat_ccopy(sloc, dloc, cname, cloc, status)
*    Description :
*     This routine copies the data object pointed to by the source
*     locator into a component of the data structure pointed to by
*     the destination locator.
*     If the object to be copied is a structure, a new structure of
*     the correct type and dimensions is created but the contents of
*     the structure are not copied.
*     A locator to the new object is returned.
*    Invocation :
*     CALL DAT_CCOPY(SLOC, DLOC, CNAME; CLOC, STATUS)
*    Parameters :
*     SLOC=CHARACTER*(DAT__SZLOC)
*           Variable containing a locator associated with the source
*           object.
*     DLOC=CHARACTER*(DAT__SZLOC)
*           Variable containing a locator associated with the data
*           structure to which the object is to be copied.
*     CNAME=CHARACTER*(DAT__SZNAM)
*           Variable containing the name of the component in the
*           structure to which the object is to be copied.
*     CLOC=CHARACTER*(DAT__SZLOC)
*           Variable containing a locator associated with the new
*           object.
*     STATUS=INTEGER
*           Variable holding the status value. If this variable is not
*           SAI__OK on input, the routine will return without action.
*           If the routine fails to complete, this variable will be
*           set to an appropriate error number.
*    Method :
*     If the source object is a primitive, do a straight copy.
*     Otherwise get the necessary quantities and create the new
*     object.
*    Authors :
*     Sid Wright  (UCL::SLW)
*    History :
*     29-Mar-1983:  Original.  (UCL::SLW)
*     05.11.1984:   Remove calls to error system.
*                   Change routine name from DAT_$CCOPY (REVAD::BDK)
*     15.04.1987:   Improve prologue layout (RLVAD::AJC)
*    Global constants :
      INCLUDE 'SAE_PAR'
      INCLUDE 'DAT_PAR'
      INCLUDE 'DAT_ERR'
*    Import :
      character*(*) sloc		! Source Locator
      character*(*) dloc		! Environment Locator
      character*(*) cname		! component to be created
*    Export :
      character*(*) cloc		! locator to new component
*    Status return :
      integer status			! Status Return
*    Local variables :
      logical struct    		! Whether object is a structure or not
      character*(DAT__SZTYP) type
      integer ndim
      integer dims(DAT__MXDIM)
      LOGICAL STATE

*-

      if (status .ne. SAI__OK) return

      call dat_struc(sloc, struct, status)
      if (status .ne. SAI__OK) return
      if (.not.struct) then
         CALL DAT_STATE( SLOC, STATE, STATUS )
         IF ( STATE ) THEN
            CALL DAT_COPY( SLOC, DLOC, CNAME, STATUS )
         ELSE
            CALL DAT_TYPE( SLOC, TYPE, STATUS )
            CALL DAT_SHAPE( SLOC, DAT__MXDIM, DIMS, NDIM, STATUS )
            CALL DAT_NEW( DLOC, CNAME, TYPE, NDIM, DIMS, STATUS )
         ENDIF
      else
         call dat_type(sloc, type, status)
         call dat_shape(sloc, DAT__MXDIM, dims, ndim, status)
         call dat_new(dloc, cname, type, ndim, dims, status)
      endif
      call dat_find(dloc, cname, cloc, status)

      end

File: libraries/hds/datccopy.c
#if HAVE_CONFIG_H
#  include <config.h>
#endif

/*+DATCCOPY.C-*/

/* Include files */
#include "ems.h"                 /* EMS error reporting routines            */

#include "hds1.h"                /* Global definitions for HDS              */
#include "rec.h"                 /* Public rec_ definitions                 */
#include "str.h"                 /* Character string import/export macros   */
#include "dat1.h"                /* Internal dat_ definitions               */
#include "dat_err.h"             /* DAT__ error code definitions            */
#include "hds.h"                 /* HDS public C interface                  */

/*=====================================*/
/* DatCcopy - Copy one structure level */
/*=====================================*/

int
datCcopy(char locator1[DAT__SZLOC],
         char locator2[DAT__SZLOC],
         char *name,
         char locator3[DAT__SZLOC],
         int  *status )
{
/*
 * This is a direct translation into C of the HDS source file dat_ccopy.f
 * See that file for all comments about the function of this routine!
 * C version - BKM 20020821 - NOT YET TESTED!
 */
 
/* Local variables */
   int structure;                             /* Structure flag */
   int state;                                 /* State flag     */
   int ndim;                                  /* Number of axes */
   HDS_PTYPE dims[DAT__MXDIM];                  /* Axes sizes     */
   char type[DAT__SZTYP+1];

/* Routine entry */
   if( !_ok(*status) )
      return *status;
     
   datStruc( locator1, &structure, status );
   if( !_ok(*status) )
      return *status;
   if( !structure ) {
      datState( locator1, &state, status );
      if ( state )
         datCopy( locator1, locator2, name, status );
      else {
         datType( locator1, type, status );
         datShape( locator1, DAT__MXDIM, dims, &ndim, status );
         datNew( locator2, name, type, ndim, dims, status );
      }
   } else {
      datType( locator1, type, status );
      datShape( locator1, DAT__MXDIM, dims, &ndim, status );
      datNew( locator2, name, type, ndim, dims, status );
   }
   datFind( locator2, name, locator3, status );

   return *status;
}
       
